### PR TITLE
[ML-DataFrame] Add _preview endpoint

### PR DESCRIPTION
**Pull request by benwtrent**
This adds a `_data_frame/transforms/_preview` endpoint that accepts a config definition and returns what the doc source's will be when it is ran. 

```
{
   data_frame_preview: [
    { /*doc source*/},
    ...
   ]
}
```

### CODE REVIEW

File: x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFramePivotRestIT.java [hunk 1]
```diff
@@ -224,6 +227,36 @@ public void testDateHistogramPivot() throws Exception {
...
+    public void testPreviewTransform() throws Exception {
+        final Request createPreviewRequest = new Request("POST", DATAFRAME_ENDPOINT + "_preview");
+
+        String config = "{"
+            + " \"source\": \"reviews\","
+            + " \"id\": \"doesnot-matter\","
+            + " \"dest\": \"doesnot-matter\",";
+
+        config += " \"pivot\": {"
+            + "   \"group_by\": ["
```
Comment (hendrikmuhs) [inline x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFramePivotRestIT.java]: the `[`, `]` need to be removed

**Comment by hendrikmuhs** [comment]
CC @peteharverson @walterra

**Comment by hendrikmuhs** [comment]
I will have a closer look on monday, due to https://github.com/elastic/elasticsearch/pull/38934 I suggest to put this on hold for now.

Good work, I like to see it in action.

File: x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFramePivotRestIT.java (hunk 1 above)
Comment (benwtrent) [reply x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFramePivotRestIT.java]: @hendrikmuhs definitely! Was waiting on your PR to be merged :D

**Comment by benwtrent** [comment]
This is blocked by #38934. will have to merge into master once that is done.

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/PreviewDataFrameTransformAction.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+        protected RequestBuilder(ElasticsearchClient client, PreviewDataFrameTransformAction action) {
+            super(client, action, new Request());
+        }
+    }
+
+    public static class Response extends ActionResponse implements ToXContentObject {
+
+        private List<Map<String, Object>> docs;
+        public static ParseField DATA_FRAME_PREVIEW = new ParseField("data_frame_preview");
```
Comment (hendrikmuhs) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/PreviewDataFrameTransformAction.java]: I wonder if transform should be somehow in the name, however it gets very long, what about just "preview"?

**Review by hendrikmuhs — COMMENTED** [review]
added some questions/comments

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/PreviewDataFrameTransformAction.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+import java.util.Map;
+import java.util.Objects;
+
+public class PreviewDataFrameTransformAction extends Action<PreviewDataFrameTransformAction.Response> {
+
+    public static final PreviewDataFrameTransformAction INSTANCE = new PreviewDataFrameTransformAction();
+    public static final String NAME = "cluster:admin/data_frame/preview";
+
+    // We need this registry for parsing out Aggregations and Searches
+    private static NamedXContentRegistry searchRegistry;
```
Comment (hendrikmuhs) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/PreviewDataFrameTransformAction.java]: you do not need this, see below

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/PreviewDataFrameTransformAction.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        private DataFrameTransformConfig config;
+
+        public Request(DataFrameTransformConfig config) {
+            this.setConfig(config);
+        }
+
+        public Request() { }
+
+        public static Request fromXContent(final XContentParser parser) throws IOException {
+            Map<String, Object> content = parser.map();
```
Comment (hendrikmuhs) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/PreviewDataFrameTransformAction.java]: with 

```
NamedXContentRegistry registry = parser.getXContentRegistry();
```

you can get the registry here and can re-use it on line 76

**Review by hendrikmuhs — APPROVED** [review]
LGTM, I suggest a small simplification

**Comment by hendrikmuhs** [comment]
still LGTM

**Pull request merged**